rgbd_depth_correction/src/depth_correction.cpp

```diff
@@ -0,0 +1,232 @@
+     * @param[in] file The file name used to the open similarly named point cloud depth correction file
+     */
+  void loadVersionOne(const YAML::Node& doc, const std::string& file);
+  /**
+     * @brief Peforms point cloud depth correction using the version one parameters and equations, then republishes the corrected cloud
+     *
+     * @param[in] cloud The point cloud to be corrected and republished
+     */
+  void correctionVersionOne(const pcl::PointCloud<pcl::PointXYZRGB>::ConstPtr &cloud);
+
+public:
+
+  /**
+     * @brief For debug testing, turns off/on the use of the depth correction coefficients to see how accurate the results are
+     *
+     * @param[in] request Empty
+     * @param[out] response Empty
+     * @return always returns true
+     */
+  bool setEnableDepth(std_srvs::Empty::Request &request, std_srvs::Empty::Response &response)
+  {
+    use_depth_exp_ = !use_depth_exp_;
+    ROS_INFO_STREAM("Using depth exponential correction: " << use_depth_exp_);
+    return true;
+  }
+
+  /**
+     * @brief On startup of nodelet, reads the YAML config file and starts the publishers/subscribers
+     */
+  virtual void onInit()
+  {
+    use_depth_exp_ = true;
+    ros::NodeHandle nh = getMTNodeHandle();
+    ros::NodeHandle priv_nh = getMTPrivateNodeHandle();
+
+    // get parameters for services, topics, and filename
+    std::string filename, filepath;
+    if(!priv_nh.getParam("filename", filename) || !priv_nh.getParam("filepath", filepath))
+    {
+      ROS_ERROR("File name and/or pathway not provided.  Closing depth correction node.");
+      return;
+    }
+
+    ROS_INFO_STREAM("Reading in yaml file " << filepath << filename);
+
+    if(!readYamlFile(filepath, filename))
+    {
+      ROS_ERROR("Error reading YAML config file.  Closing depth correction node.");
+      return;
+    }
+
+    ROS_INFO("Done reading yaml file");
+
+    pcl_pub_ = nh.advertise<pcl::PointCloud<pcl::PointXYZRGB> >("out_cloud",1);
+    pcl_sub_ = nh.subscribe("in_cloud", 1, &DepthCorrectionNodelet::pointcloudCallback, this);
+
+    depth_change_ = nh.advertiseService("change_depth_factor", &DepthCorrectionNodelet::setEnableDepth, this);
+
+  }
+};
+
+void DepthCorrectionNodelet::pointcloudCallback(const pcl::PointCloud<pcl::PointXYZRGB>::ConstPtr &cloud)
+{
+  switch(version_)
+  {
+    case 1:
+      correctionVersionOne(cloud);
+      break;
+    default:
+      ROS_ERROR_STREAM_THROTTLE(120, "Depth calibration file version does not match any known versions.  Not performing depth correction");
+      break;
+  }
+}
+
+bool DepthCorrectionNodelet::readYamlFile(const std::string &pathway, const std::string &yaml_file)
+{
+  bool rtn = true;
+  std::string file;
+  file = pathway + yaml_file + ".yaml";
+  std::ifstream fin(file.c_str());
+  if( fin.is_open())
+  {
+    try
+    {
+      YAML::Parser parser(fin);
+      YAML::Node doc;
+      parser.GetNextDocument(doc);
+
+      if(const YAML::Node *pName = doc.FindValue("version"))
```

Comment:
The syntax for yaml.cpp has changed drastically between hydro and indigo or really between ubuntu 12.04 and 14.04. For that reason, we need all yaml syntax in a separate file. Look at my latest branch for industrial_calibration. There is a file called yaml_utils.h. I have yet to write one for indigo, but it encapulates the yaml calls so that we can easily make the transition between indigo and hydro.